File: 111-minimum-depth-of-binary-tree/111-minimum-depth-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int minDepth(TreeNode* root) {
        if(root==NULL)
            return 0;
        queue <TreeNode* > q;
        q.push(root);
        int depth = 0;
        while(!q.empty())
        {
            ++depth;
            int n = q.size();
            for(int i=1;i<=n;i++)
            {
                TreeNode* head = q.front();
                q.pop();
                if(head -> left == NULL && head -> right == NULL )
                    return depth;
                if(head->left)
                    q.push(head->left);
                if(head->right)
                    q.push(head->right);
            }
        }
        return depth;
    }
};

File: 111-minimum-depth-of-binary-tree/111-minimum-depth-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int minDepth = Integer.MAX_VALUE;

    public int minDepth(TreeNode root) {
        return depth(root, 0);
    } 
    
    public int depth(TreeNode root, int depth) {
        if (root == null) {
            return 0;
        }else {
            depth++;
        }
        if (root.left == null && root.right == null && depth<minDepth) {
            minDepth = depth;
        }
        depth(root.left, depth);
        depth(root.right, depth);
        return minDepth;
        
    }
}